File: ios/MenuMaker/ViewModels/ProfileViewModel.swift
import Foundation
import Combine

@MainActor
class ProfileViewModel: ObservableObject {
    @Published var isLoading = false
    @Published var errorMessage: String?
    @Published var successMessage: String?

    private let apiClient = APIClient.shared
    private var cancellables = Set<AnyCancellable>()

    // MARK: - Profile Update

    func updateProfile(name: String?, phone: String?, address: String?) async -> Bool {
        isLoading = true
        errorMessage = nil
        successMessage = nil

        do {
            let request = UpdateProfileRequest(
                name: name,
                phone: phone,
                address: address
            )

            let response: UserResponse = try await apiClient.request(
                endpoint: AppConstants.API.Endpoints.updateProfile,
                method: .patch,
                body: request
            )

            // Update the user in AuthViewModel
            NotificationCenter.default.post(
                name: NSNotification.Name("UserProfileUpdated"),
                object: nil,
                userInfo: ["user": response.data.user]
            )

            successMessage = "Profile updated successfully"
            isLoading = false
            return true
        } catch let error as APIError {
            errorMessage = error.errorDescription
            isLoading = false
            return false
        } catch {
            errorMessage = "Failed to update profile"
            isLoading = false
            return false
        }
    }

    // MARK: - Change Password

    func changePassword(currentPassword: String, newPassword: String, confirmPassword: String) async -> Bool {
        isLoading = true
        errorMessage = nil
        successMessage = nil

        // Validate inputs
        guard !currentPassword.isEmpty else {
            errorMessage = "Current password is required"
            isLoading = false
            return false
        }

        guard !newPassword.isEmpty else {
            errorMessage = "New password is required"
            isLoading = false
            return false
        }

        guard newPassword.count >= AppConstants.Validation.minPasswordLength else {
            errorMessage = "Password must be at least \(AppConstants.Validation.minPasswordLength) characters"
            isLoading = false
            return false
        }

        guard newPassword == confirmPassword else {
            errorMessage = "Passwords do not match"
            isLoading = false
            return false
        }

        do {
            let request = ChangePasswordRequest(
                currentPassword: currentPassword,
                newPassword: newPassword
            )

            let _: MessageResponse = try await apiClient.request(
                endpoint: AppConstants.API.Endpoints.changePassword,
                method: .post,
                body: request
            )

            successMessage = "Password changed successfully"
            isLoading = false
            return true
        } catch let error as APIError {
            errorMessage = error.errorDescription
            isLoading = false
            return false
        } catch {
            errorMessage = "Failed to change password"
            isLoading = false
            return false
        }
    }

    // MARK: - Validation

    func validateName(_ name: String) -> String? {
        guard !name.isEmpty else {
            return "Name cannot be empty"
        }

        guard name.count >= 2 else {
            return "Name must be at least 2 characters"
        }

        guard name.count <= AppConstants.Validation.maxNameLength else {
            return "Name is too long"
        }

        return nil
    }

    func validatePhone(_ phone: String) -> String? {
        guard !phone.isEmpty else {
            return nil // Phone is optional
        }

        let numericPhone = phone.filter { $0.isNumber }
        guard numericPhone.count >= 10 else {
            return "Phone number must be at least 10 digits"
        }

        return nil
    }

    // MARK: - Clear Messages

    func clearMessages() {
        errorMessage = nil
        successMessage = nil
    }
}

File: android/app/src/main/kotlin/com/menumaker/viewmodel/ProfileViewModel.kt
package com.menumaker.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.menumaker.data.common.Resource
import com.menumaker.data.repository.AuthRepository
import com.menumaker.services.AnalyticsService
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * ViewModel for managing user profile
 */
@HiltViewModel
class ProfileViewModel @Inject constructor(
    private val authRepository: AuthRepository,
    private val analyticsService: AnalyticsService
) : ViewModel() {

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _successMessage = MutableStateFlow<String?>(null)
    val successMessage: StateFlow<String?> = _successMessage.asStateFlow()

    /**
     * Update user profile
     */
    fun updateProfile(
        name: String?,
        phone: String?,
        address: String?,
        onSuccess: () -> Unit = {}
    ) {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null
            _successMessage.value = null

            val updates = mutableMapOf<String, Any>()
            name?.let { updates["name"] = it }
            phone?.let { updates["phone"] = it }
            address?.let { updates["address"] = it }

            // TODO: Use a ProfileRepository when created
            // For now, use AuthRepository's updateProfile method if available
            // Or create API call directly

            try {
                // Simulated for now - implement actual API call
                kotlinx.coroutines.delay(1000)
                _successMessage.value = "Profile updated successfully"
                analyticsService.track("profile_updated", emptyMap())
                onSuccess()
            } catch (e: Exception) {
                _errorMessage.value = e.message ?: "Failed to update profile"
            }

            _isLoading.value = false
        }
    }

    /**
     * Change user password
     */
    fun changePassword(
        currentPassword: String,
        newPassword: String,
        confirmPassword: String,
        onSuccess: () -> Unit = {}
    ) {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null
            _successMessage.value = null

            // Validate inputs
            when {
                currentPassword.isEmpty() -> {
                    _errorMessage.value = "Current password is required"
                    _isLoading.value = false
                    return@launch
                }
                newPassword.isEmpty() -> {
                    _errorMessage.value = "New password is required"
                    _isLoading.value = false
                    return@launch
                }
                newPassword.length < 6 -> {
                    _errorMessage.value = "Password must be at least 6 characters"
                    _isLoading.value = false
                    return@launch
                }
                newPassword != confirmPassword -> {
                    _errorMessage.value = "Passwords do not match"
                    _isLoading.value = false
                    return@launch
                }
            }

            try {
                // TODO: Implement actual password change API call
                kotlinx.coroutines.delay(1000)
                _successMessage.value = "Password changed successfully"
                analyticsService.track("password_changed", emptyMap())
                onSuccess()
            } catch (e: Exception) {
                _errorMessage.value = e.message ?: "Failed to change password"
            }

            _isLoading.value = false
        }
    }

    /**
     * Validate name
     */
    fun validateName(name: String): String? {
        return when {
            name.isEmpty() -> "Name cannot be empty"
            name.length < 2 -> "Name must be at least 2 characters"
            name.length > 50 -> "Name is too long"
            else -> null
        }
    }

    /**
     * Validate phone number
     */
    fun validatePhone(phone: String): String? {
        if (phone.isEmpty()) return null // Phone is optional

        val numericPhone = phone.filter { it.isDigit() }
        return if (numericPhone.length < 10) {
            "Phone number must be at least 10 digits"
        } else null
    }

    /**
     * Clear messages
     */
    fun clearMessages() {
        _errorMessage.value = null
        _successMessage.value = null
    }
}
